### ISSUE TITLE
Specify demux algorithm for RTP

### ISSUE THREAD

**Issue by juberti** [rtcweb-wg/jsep#223]
Cover MID, RID, unsignaled SSRCs, and RTCP. Related to #79 and #149.

**Comment by juberti** [comment]
Suggest doing this in a new Section 5.10 on demux.

**Comment by pthatcherg** [comment]
Here's a first pass with some pseudocode.

When an RtpReceiver is added, removed, or has its parameters modified, the following is done:

for each receiver r with mid m:
  receiver_by_mid[m] = r
  for each codec c in r.getParameters().codecs:
    receiver_by_pt[codec.payloadType] = r
  for each encoding e in r.getParameters().encodings:
    receiver_by_ssrc[encoding.ssrc] = r 
    receiver_by_ssrc[encoding.rtx.ssrc] = r
    receiver_by_ssrc[encoding.fec.ssrc] = r
  for encoding e in r.getParameters().encodings:
    encoding_by_mid_and_ssrc[e.ssrc] = e
    encoding_by_mid_and_rid[m][e.rtpStreamId] = e

When an RTP packet arrives select the RTCRtpReceiver for the packet as follows:

if receiver_by_ssrc[packet.ssrc] is set to r:
  deliver packet to r
else if packet.mid is set:
  if receiver_by_mid[packet.mid] is set to r:
    if packet.payloadType is in r.getParameters().codecs[i].payloadType for some i:
      set receiver_by_ssrc[packet.ssrc] = r
      deliver packet to r
    else:
      drop packet
  else:
    drop packet
else if receiver_by_pt[packet.payloadType] is set to r:
  set receiver_by_ssrc[packet.ssrc] = r
  deliver packet to r
else:
  drop packet


And select the encoding of the RTCRtpReceiver r for the packet as follows:

if size(r.getParameters().encodings) == 1:
  deliver packet to r.getParameters().encodings[0]
else if  encoding_by_mid_and_ssrc[m][packet.ssrc] = is set to e:
  deliver packet to e
else if encoding_by_mid_and_rid[m][packet.rtpStreamId] or encoding_by_mid_and_rid[m][packet.repairedRtpStreamId] is set to e:
  set encoding_by_ssrc[packet.ssrc] = e
  deliver packet to e
 

TODO: RTCP

**Comment by pthatcherg** [comment]
Actually, now that I think about it, we don't need to specify how to handle RID at all.  We could just way "receive all the source RTP streams that come in and use RtpStreamID and RepairedRtpStreamID to correlate source RTP streams and repair RTP streams".  In other words, the RIDs don't need to be signalled to the receiver.  

If I'm correct about this, we can have a simplified version:

When an RtpReceiver is added, removed, or has its parameters modified, the following is done:

for each receiver r with mid m:
  receiver_by_mid[m] = r
  for each codec c in r.getParameters().codecs:
    receiver_by_pt[codec.payloadType] = r
  for each encoding e in r.getParameters().encodings:
    receiver_by_ssrc[encoding.ssrc] = r 
    receiver_by_ssrc[encoding.rtx.ssrc] = r
    receiver_by_ssrc[encoding.fec.ssrc] = r

When an RTP packet arrives select the RTCRtpReceiver for the packet as follows:

if receiver_by_ssrc[packet.ssrc] is set to r:
  deliver packet to r
else if packet.mid is set:
  if receiver_by_mid[packet.mid] is set to r:
    if packet.payloadType is in r.getParameters().codecs[i].payloadType for some i:
      set receiver_by_ssrc[packet.ssrc] = r
      deliver packet to r
    else:
      drop packet
  else:
    drop packet
else if receiver_by_pt[packet.payloadType] is set to r:
  set receiver_by_ssrc[packet.ssrc] = r
  deliver packet to r
else:
  drop packet

**Comment by gloinul** [comment]
We had discussion during IETF that there is a need for a RTP media plane specification related to simulcast and the different types of RTP middleboxes to clarify where the various SDES items need to occur and which types of SSRC, MID and RID switching behavior that can occur on the various legs. This will clearly spark discussion as there was diverse views between Emil Ivov, Bernard Aboba, and Jonathan Lennox. Varun Singh had observed I think 4 different behaviours.

**Comment by fluffy** [comment]
Need to not forget the case where there are multiple ports that help disambiguate.

**Comment by pthatcherg** [comment]
For the case of multiple ports: I think all we need to do is point out that the algorithm applies to each of the RtpReceivers listening on a given port.  In other words, demux is per-port.  And obviously if you have only one RtpReceiver per port, there's nothing to do except check that the PT is known, which I think the algorithm I posted covers.

**Comment by ekr** [comment]
This looks reasonable except that the style is a bit inconsistent. specifically:

for each encoding e in r.getParameters().encodings:
    receiver_by_ssrc[encoding.ssrc] = r

**Comment by pthatcherg** [comment]
Here's some possible text, with slightly tweaked pseudo-code to match:


When an RTP packet is received by a transport (typically a DtlsTransport), that packet MUST be routed to the correct RtpReceiver.  For each transport, the following steps MUST be followed to prepare to route packets:

1. Construct a table mapping MID to RtpReceiver.  For each RtpReceiver configured to receive from this transport, let the key be the RtpReceiver's MID, and the value be the RtpReceiver.   
2. Construct a table mapping SSRC to RtpReceiver.  For each RtpReceiver configured to receive from this transport, and for each SSRC that RtpReceiver is configured to receive, let the key be the SSRC and the value be the RtpReceiver. 
3. Construct a table mapping Payload Type to RtpReceiver.  For each RtpReceiver configured to receive from this transport, and for each Payload Type that RtpReceiver is configured to receive, let the key be the Payload Type and the value be the RtpReceiver.

For each RTP packet received, the following steps MUST be followed to route the packet:

1. If the packet has a MID and that MID is not in the table mapping MID to RtpReceiver, drop the packet and stop.
2. If the packet has a MID and that MID is in the table mapping MID to RtpReceiver, update the SSRC mapping table to include an entry mapping the packet's SSRC to the RtpReceiver. 
3. If the packet's SSRC is in the SSRC mapping table, route the packet to the mapped RtpReceiver and stop.
4. If the packet's Payload Type is in the Payload Type table, update the the SSRC mapping table to include an entry mapping the packet’s SSRC to the RtpReceiver. Deliver the packet to the RtpReceiver and stop.
5. Drop the packet.

Note:  When designing these steps, a troubling possible use case was if the MIDs were lost in signaling, but still sent in RTP packets.  But a packet would only be sent with MIDs if an answer were sent with MIDs in it.  If there are no MIDs in the answer, then MIDs would not be present in the RTP.  And if MIDs are in the answer, they weren't lost from the signaling.  In other words, this case will not happen.


This would correspond to pseudocode something like this:

```
// Setup
receiver_by_mid = {}
receiver_by_ssrc = {}
receiver_by_pt = {}
for tr in PeerConnection.getTransceivers():
  if tr.receiver.transport == transport:
    receiver_by_mid[tr.mid] = tr.receiver
    for codec in tr.receiver.getParameters().codecs:
      receiver_by_pt[codec.payloadType] = tr.receiver
    for encoding in tr.receiver.getParameters().encodings:
      receiver_by_ssrc[encoding.ssrc] = tr.receiver
      receiver_by_ssrc[encoding.rtx.ssrc] = tr.receiver
      receiver_by_ssrc[encoding.fec.ssrc] = tr.receiver

// Per packet
if packet.mid is set:
  if receiver_by_mid[packet.mid] is not set:
    drop packet
    return
  receiver_by_ssrc[packet.ssrc] = receiver_by_mid[packet.mid]
if receiver_by_ssrc[packet.ssrc] is set:
  deliver packet to receiver_by_ssrc[packet.ssrc]
  return
if receiver_by_pt[packet.payloadType] and packet.mid is not set:
  receiver_by_ssrc[packet.ssrc] = receiver_by_pt[packet.payloadType]
  deliver packet to receiver_by_pt[packet.payloadType]
  return
drop packet
```

**Comment by pthatcherg** [comment]
PR: https://github.com/rtcweb-wg/jsep/pull/320

**Comment by juberti** [comment]
Resolved by #320
